File: python/phonebook.py
from typing import Dict, List, Optional


class PhoneBook:
    """
    Created by leon on 1/23/18.
    Made WAY better by kristofer 6/16/20
    Python version for beginner coders
    """

    def __init__(self, phonebook_dict: Optional[Dict[str, List[str]]] = None):
        """
        Constructor for PhoneBook
        :param phonebook_dict: Optional dictionary to initialize the phonebook with
        """
        if phonebook_dict is None:
            self.phonebook = {}
        else:
            self.phonebook = phonebook_dict



    def add(self, name: str, phone_number: str) -> None:
        """
        Add a phone number for a contact
        :param name: Contact name
        :param phone_number: Phone number to add
        """
        if name not in self.phonebook:
            self.phonebook[name]=[phone_number]
        else:
            self.phonebook[name].append(phone_number)


    def add_all(self, name: str, *phone_numbers: str) -> None:
        """
        Add multiple phone numbers for a contact
        :param name: Contact name
        :param phone_numbers: Variable number of phone numbers to add
        """
        if name not in self.phonebook:
            self.phonebook[name]=[]
            for num in phone_numbers:
                self.add(name, num)
        else:
            for num in phone_numbers:
                self.add(name, num)        

    def remove(self, name: str) -> None:
        """
        Remove a contact from the phonebook
        :param name: Contact name to remove
        """
        if name not in self.phonebook:
            return None
        del self.phonebook[name]
        



    def has_entry(self, name: str, phone_number: str = None) -> bool:
        """
        Check if a contact exists, optionally with a specific phone number
        :param name: Contact name to check
        :param phone_number: Optional phone number to check
        :return: True if contact exists (with phone number if specified), False otherwise
        """
        if name not in self.phonebook:
            return False
        if phone_number not in self.phonebook[name]:
            return False
        return True
            


    def lookup(self, name: str) -> List[str]:
        """
        Look up all phone numbers for a contact
        :param name: Contact name to look up
        :return: List of phone numbers for the contact
        """
        if name not in self.phonebook:
            return None
        return self.phonebook[name]

    def reverse_lookup(self, phone_number: str) -> str:
        """
        Find the contact name for a given phone number
        :param phone_number: Phone number to look up
        :return: Contact name associated with the phone number
        """
        for key in self.phonebook:
            if phone_number in self.phonebook[key]:
                return key

    def get_all_contact_names(self) -> List[str]:
        """
        Get all contact names in the phonebook
        :return: List of all contact names
        """
        return list(self.phonebook.keys())

    def get_map(self) -> Dict[str, List[str]]:
        """
        Get the underlying dictionary representation of the phonebook
        :return: Dictionary mapping names to lists of phone numbers
        """
        return self.phonebook

File: java/src/main/java/com/zipcodewilmington/phonebook/PhoneBook.java
package com.zipcodewilmington.phonebook;

import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
//import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Created by leon on 1/23/18.
 * Made WAY better by kristofer 6/16/20
 */
public class PhoneBook {

    private final Map<String, List<String>> phonebook;

    public PhoneBook(Map<String, List<String>> map) {
        this.phonebook = map;
    }

    public PhoneBook() {
        phonebook = new HashMap<>() ;
    }

    public void add(String name, String phoneNumber) {
         List<String> temp;
        if(phonebook.containsKey(name)){
            temp = phonebook.get(name);
        }else{
            temp = new ArrayList<>();
        }
        temp.add(phoneNumber);
        phonebook.put(name, temp);
    }

    public void addAll(String name, String... phoneNumbers) {
        for(String num : phoneNumbers){
           add(name, num);
        }

        
    }

    public void remove(String name) {
        phonebook.remove(name);
    }

    public Boolean hasEntry(String name, String phonenum) {

        if(phonenum==null){
            return phonebook.containsKey(name);
        }

        return phonebook.get(name).contains(phonenum);
    }
    public Boolean hasEntry(String name){
        return hasEntry(name, null);
    }

    public List<String> lookup(String name) {
        return phonebook.get(name);
    }

    public String reverseLookup(String phoneNumber)  {

        for(String name : phonebook.keySet()){
            for(String num:phonebook.get(name)){
                if(num.equals(phoneNumber))
                    return name;
            }
        }
        return null;
    }

    public List<String> getAllContactNames() {
        // return new ArrayList<>(phonebook.keySet());
        List<String> temp = new ArrayList<>();
        for(String n : phonebook.keySet())
            temp.add(n);
        return temp;    
    }

    public Map<String, List<String>> getMap() {
        return phonebook;
    }
}
